File: code_translation/exp1_code_translation_security/translated_solutions_o3/Question1/Go/1012.go
package main

import (
	"fmt"
	"log"

	"github.com/fernet/fernet-go"
)

// ---------------------------------------------------------------------------
// Generate a single global key (equivalent to: key = Fernet.generate_key()).
// ---------------------------------------------------------------------------
var key *fernet.Key

func init() {
	var k fernet.Key
	if err := k.Generate(); err != nil {
		log.Fatalf("unable to generate fernet key: %v", err)
	}
	key = &k
}

// ---------------------------------------------------------------------------
// encryptText :  string -> string
//   Mirrors  f.encrypt(bytes(plain_text, "utf-8")).decode()
// ---------------------------------------------------------------------------
func encryptText(plainText string) string {
	// EncryptAndSign returns a base64-encoded token.
	token := fernet.EncryptAndSign([]byte(plainText), key)
	return string(token)
}

// ---------------------------------------------------------------------------
// decryptText :  string -> (string, error)
//   Mirrors  f.decrypt(token).decode()
//   TTL is set to 0 => no expiration check (same as default in Python snippet).
// ---------------------------------------------------------------------------
func decryptText(encryptedText string) (string, error) {
	plaintext := fernet.VerifyAndDecrypt([]byte(encryptedText), 0, []*fernet.Key{key})
	if plaintext == nil {
		return "", fmt.Errorf("decryption failed or token invalid")
	}
	return string(plaintext), nil
}

// ---------------------------------------------------------------------------
// Simple demo
// ---------------------------------------------------------------------------
func main() {
	msg := "Hello from Go + Fernet!"

	token := encryptText(msg)
	fmt.Println("Encrypted token:", token)

	plain, err := decryptText(token)
	if err != nil {
		log.Fatalf("decryptText error: %v", err)
	}
	fmt.Println("Decrypted text :", plain)
}

File: code_translation/exp1_code_translation_security/translated_solutions_o3/Question1/Rust/1012.rs
// Cargo.toml (for reference only)
// [dependencies]
// fernet    = "0.2"      # Fernet implementation for Rust
// once_cell = "1"        # For a lazily-initialized static key

use once_cell::sync::Lazy;
use fernet::{Fernet, DecryptError};

// Lazily generate a single global key, exactly like the Python snippet did.
static KEY: Lazy<String> = Lazy::new(|| Fernet::generate_key());

/// Encrypt a UTF-8 text string and return the resulting Fernet token
/// (base-64-encoded, just like the Python `cryptography` library).
pub fn encrypt_text(plain_text: &str) -> String {
    // `encrypt` returns a `String`, so no additional conversion is needed.
    Fernet::encrypt(&KEY, plain_text)
}

/// Decrypt a Fernet token back into a UTF-8 `String`.
pub fn decrypt_text(encrypted_text: &str) -> Result<String, DecryptError> {
    Fernet::decrypt(&KEY, encrypted_text)
}

// --- Example usage ---------------------------------------------------------
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip() {
        let msg  = "Hello, Rust!";
        let enc  = encrypt_text(msg);
        let dec  = decrypt_text(&enc).expect("decryption failed");
        assert_eq!(msg, dec);
    }
}